lib/exgboost/training/callback.ex

```diff
@@ -46,30 +46,35 @@ defmodule EXGBoost.Training.Callback do
   @enforce_keys [:event, :fun]
   defstruct [:event, :fun, :name, :init_state]
 
-  @doc """
-  Factory for a new callback without an initial state. See `EXGBoost.Callback.new/4` for more details.
-  """
-  @spec new(
-          event :: :before_training | :after_training | :before_iteration | :after_iteration,
-          fun :: (State.t() -> {:cont, State.t()} | {:halt, State.t()})
-        ) :: Callback.t()
-  def new(event, fun) do
-    new(event, fun, nil, %{})
-  end
+  @type event :: :before_training | :after_training | :before_iteration | :after_iteration
+  @type fun :: (State.t() -> State.t())
+
+  @valid_events [:before_training, :after_training, :before_iteration, :after_iteration]
 
   @doc """
   Factory for a new callback with an initial state.
   """
-  @spec new(
-          event :: :before_training | :after_training | :before_iteration | :after_iteration,
-          fun :: (State.t() -> {:cont, State.t()} | {:halt, State.t()}),
-          name :: atom(),
-          init_state :: map()
-        ) :: Callback.t()
-  def new(event, fun, name, %{} = init_state)
-      when event in [:before_training, :after_training, :before_iteration, :after_iteration] and
-             is_atom(name) do
+  @spec new(event :: event(), fun :: fun(), name :: atom(), init_state :: any()) :: Callback.t()
+  def new(event, fun, name, init_state \\ %{})
+      when event in @valid_events and is_function(fun, 1) and is_atom(name) and not is_nil(name) do
     %__MODULE__{event: event, fun: fun, name: name, init_state: init_state}
+    |> validate!()
+  end
+
+  def validate!(%__MODULE__{} = callback) do
```

Comment:
My only hangup here is that the only reason we need a name is to look up associated `meta_vars`, so there should still be the ability to create a callback that has no state and no name

Reply:
Oh I totally forgot about this part. Sorry I meant to call it out.

My thinking was this: we can't use `name` as the key in `meta_vars` unless all names are unique. Multiple callbacks with `name: nil` break that. Since I couldn't find an instance where `name: nil` was actually used, I figured I'd just make it required.

If we _do_ want to allow `name: nil`, then we need to alter the creation of `meta_vars` to account for duplicates. Come to think of it, we need to do that anyway.

Reply:
My thought was that you just check for `nil` name and dont even look at meta vars if its `nil` and instead just run the func. And then the `name`s would have to be unique still

Reply:
I addressed the duplicate name issue and wrote a test.

[f1977bf](https://github.com/acalejos/exgboost/pull/35/commits/f1977bf9190816eb237505471befb9291b5c675b)

This doesn't address the question about how to handle `name: nil`, but it prevents us clobbering callbacks when we build `meta_vars`. We can easily filter `name: nil` callbacks out or handle them with some other convention.

Thoughts?

Reply:
FYI, I think the other way to handle this is to do something like:

```elixir
meta_vars =
  callbacks
  |> Enum.sort_by(& &1.name) # for determinism
  |> Enum.with_index()
  |> Map.new(&{&1.name, &1.init_state})
```

That side-steps the name issue altogether. But then if duplicate names are present, it's not possible to provide an error message like "Callback X did the wrong thing.".